New Feature: Schedule featured articles #13596

eshiol opened the pull request:
Pull Request for Issue #13596.

### Summary of Changes
Added two fields in the article editor `featured_up` (Start Featured) and `featured_down` (End Featured)
![image](https://user-images.githubusercontent.com/12718836/27852039-d6711708-615d-11e7-8e53-1b8494c12532.png)
An article is featured if `featured` is yes and (`featured_up` is null or is less than or is equal to now) and (`featured_down` is null or is greater than or is equal to now)

### Testing Instructions
create 5 articles:
1) not featured
2) featured, featured_up = null, featured_down = null
3) featured, featured_up = tomorrow, featured_down = null
4) featured, featured_up = null, featured_down = yesterday
5) featured, featured_up = yesterday, featured_down = tomorrow

Create an `Articles » Featured Articles` menu item
Create a `News Flash` module and set the option `Featured Articles` to `Only show Featured Articles`.

### Expected result
Only the articles 2 and 5 will be shown in the `Articles » Featured Articles` menu item and in the `News Flash` module.

### Actual result

### Documentation Changes Required

laoneo committed:
[namespace] Move legacy controller classes to namespace (#13683)

* Move JControllerLegacy to namespace

* Change content controller to use the new Base controller class

* Adapting the test as JControllerLegacy has a different class name

* New line at the end of the file

* Activating tests

* Reset to the original state

* Rename Base to Controller

* Change name to new class name Controller

* Move to Cms namespace

* Load the Cms namespace trough composer

* Fix the test again

* Composer dump with latest version

* Move to Joomla only namespace in composer.json

* Move JControllerAdmin to namespace

* Happy travis

* Add missing alias

laoneo committed:
[namespace] Move JViewLegacy to namespace (#13697)

* Move JViewLegacy to namespace

* Add alias for JViewLegacy on JLoader

* Move the categories views to namespace

laoneo committed:
[namespace] Move JModel* to namespace (#13704)

* Move JModelLegacy to namespace

* Move JModelList to namespace

* Move JModelAdmin to namespace

* Move JModelForm to namespace

* Also use the proper namespace for exceptions

* Update the class alias list

* Namespace stdClass

* Fixe some simple tests

* Auch

* Extend from new model class

* Remove tests which rely on the hash cache

* CS fixes

wilsonge committed:
Merge branch 'staging' into 3.8.x

laoneo committed:
Fix namespace for exceptions (#13722)

laoneo committed:
Fix core namespace (#13720)

laoneo committed:
[namespace] Move JTable* to namespaces (#13765)

* Fix namespace for exceptions

* Move JTable to namespace

* Move JTableObserver to namespace

* Move JTableInterface to namespace

* Move JTableNested to namespace

* Move JTableObserverContenthistory to namespace

* Move JTableObserverTags to namespace

* Fix code style and unit test

* Remove package information in docblock

* Package is not required anymore in docblock

* Revert invalid composer commit

* Move JTableAsset to namespace

* Move JTableExtension to namespace

* Move JTableLanguage to namespace

* Move JTableUpdate to namespace

* Move JTableUpdatesite to namespace

* Move JTableUser to namespace

* Move JTableUsergroup to namespace

* Move JTableViewlevel to namespace

* Update doc block, remove package declaration

* Fix intendion cs error

wilsonge committed:
Merge branch 'staging' into 3.8-dev

wilsonge committed:
Merge branch 'staging' into 3.8-dev

wilsonge committed:
Merge branch 'staging' into 3.8-dev

laoneo committed:
[namespacce] Namespace JAccess* (#13952)

* Move JAccess to namespace

* Fix some minor stuff and delete JRule and JRules

wilsonge committed:
Merge branch 'staging' into 3.8-dev

wilsonge committed:
Merge branch 'staging' into 3.8-dev

wilsonge committed:
Merge branch 'staging' into 3.8-dev

wilsonge committed:
Merge branch 'staging' into 3.8-dev

laoneo committed:
Move JAuthenticationHelper to namespace (#14155)

wilsonge committed:
Merge branch 'staging' into 3.8-dev

wilsonge committed:
Merge branch '3.8-dev' of github.com:joomla/joomla-cms into 3.8-dev

wilsonge committed:
Fix code sniffer

zero-24 committed:
move 3.8 branch to 3.8 in the version file (#14167)

zero-24 committed:
Update classmap.php (#14165)

laoneo committed:
Move JHelp to namespace (#14232)

laoneo committed:
Move JLanguage* to namespace (#14410)

wilsonge committed:
Merge branch 'staging' into 3.8-dev

wilsonge committed:
Merge branch 'staging' into 3.8-dev

laoneo committed:
Support alias overrides and overrides for namespaced classes (#14822)

* Add override plugin

* Support overrides of namespaced classes

* Better function name

* Move override plugin to rep

* Revert to none namespaced version as on staging

* Load the Joomla namespace trough JLoader

* JLoader alis load class support

* CS

* Only load the alias when the original class exists

* One more space, outch

* Add unit tests

* Add note about use of composer

laoneo committed:
Move JCaptcha to namespace (#14225)

* Move JCaptcha to namespace

* Change CMS classes as well

* Adapted intendation

laoneo committed:
Move JErrorPage to namespace (#14231)

* Move JErrorPage to namespace

* Doc block comment should be inline with the rest

* Move class to Joomla\Cms\Exception\ExceptionHandler

* Correct class name

laoneo committed:
Move JEditor to namespace (#14230)

laoneo committed:
Move JComponent* to namespace (#14229)

* Change JComponent* classes to namespace

* Adapted intendation

* Check for new interface

* Intendation

wilsonge committed:
Fix fatal error

wilsonge committed:
Merge branch 'staging' into 3.8-dev

wilsonge committed:
Fix fatal errors from #14229

wilsonge committed:
Merge branch 'staging' into 3.8-dev

Spudley committed:
Replace dirname(__FILE__) with __DIR__ (#14269)

* Replace dirname(__FILE__) with __DIR__

* Revert restore.php change per request.

* do'h my IDE converted tabs to spaces.

* Revert installation/index.php per request.

laoneo committed:
Move JComponentRecord to namespace (#14996)

* Move JComponentRecord to namespace

* licens file should be license.txt

* Fix invalid merges

* doc

* new doc block

brianteeman committed:
change from cms to CMS

brianteeman committed:
change folder name

brianteeman committed:
renamed back to cms it was the wrong folder

brianteeman committed:
renamed back to cms

brianteeman committed:
renamed the correct folder

laoneo committed:
Move Application classes to namespace and fix namespace move to CMS (#15085)

wilsonge committed:
Merge branch 'staging' into 3.8-dev

zero-24 committed:
use Development and not dev as Dev Status

wilsonge committed:
Merge branch 'staging' into 3.8-dev

laoneo committed:
Move helper classes to namespace (#15070)

* Move helper classes to namespace

* Move AuthenticationHelper to helper namespace

* Some cleanup

* add missing media helper entry

* intendation

* revert

laoneo committed:
Move CMS tables to namespace (#15000)

* Move CMS tables to namespace

* CS and invalid doc blocks

* Fix

* Use table getInstance

* doc

* upper case CMS

* fix

wilsonge committed:
Merge branch '3.8-dev' of github.com:joomla/joomla-cms into 3.8-dev

laoneo committed:
Move JLayout* to namespace (#15209)

brianteeman committed:
[3.8 dev] namespace CMS (#15224)

* rename CMSApplication

* rename

* rename

laoneo committed:
Move JLanguage* to namespace (#15216)

* Move JLanguage* to namespace

* Change JFactory to namespaced language class

* Namespace core classes

laoneo committed:
Namespace JModuleHelper (#15242)

laoneo committed:
Move JLibraryHelper to namespace (#15239)

* Move JLibraryHelper to namespace

* Fix stdclass namespace error

laoneo committed:
Move JResponseJson to namespace (#15248)

laoneo committed:
Move JSearchHelper to namespace (#15246)

* Move JSearchHelper to namespace

* CS

* Revert cleanup

* NS JFactory call

laoneo committed:
Move JPlugin to namespace (#15267)

brianteeman committed:
Rename folder to Wrapper (#15262)

wilsonge committed:
Merge branch 'staging' into 3.8-dev

Commit:
Revert "manuell revert (#15208)"

This reverts commit 68bcb875f864a726d47d15e946d34d7c4b7048a0.

Commit:
Revert "just remove the UI (#14919)"

This reverts commit 0c48d9220964dd9b23cc18d39ee162cceb073725.

Commit:
Merge branch 'staging' of github.com:joomla/joomla-cms into 3.8-dev

Commit:
Deprecate Bootstrap Affix as this doesn't exist in BS4 (#15383)

* Deprecation

* Added Typehead

* Update bootstrap.php

laoneo committed:
Namespace menu (#15284)

* Move files

* Change file

* Change file

* Change code

* Abstract menu name and classmap

* Pedantic whitespace change

laoneo committed:
Namespace pagination (#15405)

* Namespace pagination

* Put versions in inline deprecations

laoneo committed:
Namespace pathway (#15418)

laoneo committed:
Namespace schema (#15419)

* Namespace schema

* CS

wilsonge committed:
Merge branch 'staging' into 3.8-dev

laoneo committed:
Namespace ucm (#15437)

* Namespace ucm

* cs

wilsonge committed:
Merge branch 'staging' into 3.8-dev

laoneo committed:
Some namespacing fixes for captcha (#15484)

laoneo committed:
Namespace version (#15483)

Commit:
Merge branch 'staging' of github.com:joomla/joomla-cms into 3.8-dev

laoneo committed:
Namespace authentication (#15627)

laoneo committed:
Namespace association (#15625)

wilsonge committed:
Merge branch 'staging' into 3.8-dev

brianteeman committed:
deprecate hathor specific code (#15646)

Add a comment to mark hathor specific code as deprectaed for j4

Commit:
Update scripts with new version class file path

Commit:
Merge branch 'staging' of github.com:joomla/joomla-cms into 3.8-dev

Commit:
Fix stdClass call

Commit:
Fix namespace references

Commit:
Restore language strings

This reverts 652d1f898001be621ba078f7cb57f52021e60ac8 and partially reverts 1e06884ec38ee2a32eff70103440bacc48ff5d76

laoneo committed:
Namespace toolbar (#15482)

* Namespace toolbar

* Load the alias always

* Fix JButton alias

* Import fix from #15484

* Revert

* Move to ns

* Move to ns

laoneo committed:
Register JToolBar correctly (#15923)

Commit:
Deleted files for 3.8 (#15965)

Commit:
Don't remove directory (Fix #15975)

wilsonge committed:
Fix return type

laoneo committed:
Move installer libraries to namespace (#15884)

* Move installer libraries to namespace

* CS

laoneo committed:
Namespace router (#15438)

* Namespace router

* cs

* Merge remote-tracking branch 'remotes/upstream/3.8-dev' into j3/namespace/router

# Conflicts:
#	libraries/classmap.php

* Move constants to legacy file

* Adding constants to import.php as well

laoneo committed:
Move browser to namespace (#15907)

* Move browser to namespace

* Fix failing include in browser unit test

* Change to a class which exists

* argh

wilsonge committed:
Merge branch 'staging' into 3.8-dev

wilsonge committed:
Fix merge conflict

wilsonge committed:
Add missing import

laoneo committed:
Move JDate to namespace (#15900)

* Move JDate to namespace

* Change JFactory to use the namespaced date class

* Load the cms before JError to have the namespaced date class available

* Merge remote-tracking branch 'remotes/upstream/3.8-dev' into j3/namesapce/date

# Conflicts:
#	libraries/classmap.php

Commit:
Merge branch 'staging' of github.com:joomla/joomla-cms into 3.8-dev

Commit:
Catch up deleted files

zero-24 committed:
Support collection of custom stats in stats frontend module (#15138)

* catch exeptions in geLanguages

* Revert "catch exeptions in geLanguages"

This reverts commit 8072e5833d8a857f556298b2515b9b4056a90abb.

* implement onGetStats also to the frontend

* fix missing $i++

* thanks @astridx

alikon committed:
[com_content] - add categories filter to archived view (#15184)

* [com_content] - archived view add categories filter

archived view add categories filter

* [com_content] - archived view add categories filter

archived view add categories filter

* move catid filter to 1st tab

move catid filter to 1st tab

* move to 1st tab

move to 1st tab

* drone fix

drone fix i hope

* avoid confusion 

avoid confusion in implode()

Commit:
Fix namespace references

Commit:
Merge branch '3.8-dev' of github.com:joomla/joomla-cms into 3.8-dev

laoneo committed:
Move logger to namespace (#16242)

* Move logger to namespace

* Set up the classloaders before everything else

mbabker committed:
Fix namespace

mbabker committed:
Merge branch '3.8-dev' of github.com:joomla/joomla-cms into 3.8-dev

mbabker committed:
Fix namespace

mbabker committed:
Merge branch 'staging' of github.com:joomla/joomla-cms into 3.8-dev

mbabker committed:
Log package in deleted files

mbabker committed:
Use namespaced class names in test helper

laoneo committed:
Namespace categories (#16209)

laoneo committed:
Namespace profiler (#16194)

* Namespace profiler

* Merge remote-tracking branch 'remotes/upstream/3.8-dev' into j3/namespace/profiler

# Conflicts:
#	libraries/classmap.php

* Correct namespace

laoneo committed:
Namespace uri (#16195)

laoneo committed:
Namespace cache (#16298)

* Namespace cache

* Use namespaced classes

* Support old replace

* Add missing class alias

* Add missing class alias

* Fix class aliases

* Use ns class in JFactory

laoneo committed:
Namespace user (#16214)

* Namespace user

* Namespace user

* Use ns'd hash class

* Merge remote-tracking branch 'remotes/upstream/3.8-dev' into j3/namespace/user

# Conflicts:
#	libraries/classmap.php
#	libraries/src/Joomla/CMS/User/UserHelper.php

laoneo committed:
Namespace session (#16302)

* Namespace session

* Namespace session

laoneo committed:
Fix session not class found error (#16308)

mbabker committed:
Sync deleted files

laoneo committed:
Fix some session stuff (#16310)

* Fix various core classes import

* Use old path

mbabker committed:
Remove trailing spaces

laoneo committed:
Use the correct classpath for CacheExceptionInterface (#16327)

mbabker committed:
Stub generator script (#16291)

mbabker committed:
Fix calculation of the adapter type

laoneo committed:
Use the JLogEntry class (#16342)

laoneo committed:
Change deprecate version from 4.0 to 5.0 (#16353)

fastslack committed:
Namespace filter (#16359)

* Namespace filter

* Fix OutputFilter classname

mbabker committed:
Update deleted files

laoneo committed:
[3.8] Load the alias class when the original is loaded (#16361)

* Load the alias when a class is loaded in JLoader

* Revert JLog workaround

* Load the inspector on setUp

* Load the inspector on setUp

* Restore classAliasesInverse

laoneo committed:
Namespace legacy tables (#16306)

* Namespace legacy tables

* cs

* Merge remote-tracking branch 'remotes/upstream/3.8-dev' into j3/namespace/table

# Conflicts:
#	libraries/classmap.php

* use a class which exists

laoneo committed:
[3.8] Namespace form, rules and a field (#16396)

* Namespace form

* Fix ut

* doing bc

mbabker committed:
Merge branch 'staging' of github.com:joomla/joomla-cms into 3.8-dev

laoneo committed:
Move microdata to namespace (#16212)

mbabker committed:
Deprecate wrappers (#16446)

mbabker committed:
Sync deleted files

mbabker committed:
Rename class to match 4.0 and avoid another rename

laoneo committed:
Deprecate JGrid (#16211)

* Deprecate JGrid

* CS

wilsonge committed:
Merge branch 'staging' into 3.8-dev

mbabker committed:
Fix view class name refs

wilsonge committed:
Merge branch 'staging' into 3.8-dev

wilsonge committed:
Fix classmap version to match #16446

mbabker committed:
Merge branch 'staging' of github.com:joomla/joomla-cms into 3.8-dev

laoneo committed:
Namespace JFactory (#16636)

* Namespace JFactory

* JFactory is autoloaded, remove the obsolete tests

joomdonation committed:
Fix classname conflict (#16638)

laoneo committed:
Don't delete the legacy application folder (#16687)

laoneo committed:
Namespace mail (#16692)

mbabker committed:
Fix failing JMail test

mbabker committed:
Merge branch 'staging' of github.com:joomla/joomla-cms into 3.8-dev

mbabker committed:
Catch up deleted files list

laoneo committed:
Namespace client (#16634)

* Namespace client

* spelling

* change dep version

mbabker committed:
Classmap deprecated class stubs (#16760)

mbabker committed:
Remove tests for default behavior as these paths are no longer loaded by default by the library. Also, the FormHelper should really have its own test class...

mbabker committed:
Merge branch '3.8-dev' of github.com:joomla/joomla-cms into 3.8-dev

mbabker committed:
Undeprecate and namespace unsupported session store exception

phproberto committed:
[fix] table exceptions after namespacing (#16793)

laoneo committed:
Namespace updater (#16822)

* Namespace updater

* Uncomment

* Not used

* Remove not needed imports

* cosmetic

laoneo committed:
Introduce delegating psr 3 logger (#16818)

* Introduce delegating psr 3 logger

* nl

* correct since

laoneo committed:
Deprecate route wrapper (#16829)

wilsonge committed:
Merge branch 'staging' into 3.8-dev

wilsonge committed:
Merge pull request #16784 from mbabker/session-exception

[3.8] Undeprecate and namespace unsupported session store exception

mbabker committed:
Catch up deleted files

mbabker committed:
Rename file to use camel casing

mbabker committed:
Correct doc block reference in Categories, use namespaced factory

mbabker committed:
Rename Exception class to be a bit more clear, fix reference to root Exception class

mbabker committed:
Missed file from last commit

mbabker committed:
UCM is an acronym, make it all caps

laoneo committed:
Namespace crypt (#16825)

* Namespace crypt

* Rename to Key

* Rename interface

* CS

* single p

* only key

* Fix merge conflicts errors

laoneo committed:
Use the Archive package from framework and deprecate JArchive (#14157)

* Add archive package

* Deprecate JArchive

* Deprecate JArchive

* Use the Archive class instead of JArchive

* Wrong indents

* Wrong indents second part

* Setting tmp path correct

* adapt the extract dir based on file extension

* remove debug code

* Use the latest code from the archive class till version 1.1.5 is released

* Revert workaround

* Adapt installer

* Use the archive

* Deprecate the adapters

* CS

* CS

* CS

* CS

* Use version 1.1.5

* zero 24 cs

laoneo committed:
CS on classmap file (#16898)

laoneo committed:
Namespace CMS fields

wilsonge committed:
Merge pull request #16899 from Digital-Peak/j3/namesapce/cmsfields

Namespace CMS fields

mattiaverga committed:
Add css class option to image element of menu items (#16456)

* Add css style option to menu image

* Add css image for all menu item types

* Language typo

* Fix indentation

* Fix language strings

* Use tabs indentation in *.xml

* cs

* Capitalize 'S'

wilsonge committed:
Merge branch 'staging' into 3.8-dev

wilsonge committed:
No need for front slash

mbabker committed:
Catch up deleted files

laoneo committed:
Namespace punicode (#16931)

* Namespace punicode class

* Deprecate wrappers

* Move to string namespace

laoneo committed:
Namespace utility classes (#16934)

* Namespace utility classes

* Correct name of the folder

* Move register function to the top of the file

eshiol committed:
Schedule featured articles #13596

eshiol committed:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into Schedule-featured-articles-#13596

eshiol committed:
fixed query

tonypartridge commented:
Have you tested the modules view which featured are shown only?

eshiol committed:
fixed filter.featured == 'only'

fixed on the modules view which featured are shown only

eshiol committed:
Code Style fixes for components/com_content/models/articles.php

eshiol commented:
I tested and fixed the code. thx @tonypartridge

alikon commented:
ummmhh
i guess that `featured_up` ,`featured_down` should be added to  the `#__content_frontpage` table

brianteeman committed:
codestyle on the xml

brianteeman commented on administrator/language/en-GB/en-GB.com_content.ini:
```diff
@@ -58,6 +58,10 @@ COM_CONTENT_FIELD_CREATED_BY_LABEL="Created By"
 COM_CONTENT_FIELD_CREATED_DESC="Created date."
 COM_CONTENT_FIELD_CREATED_LABEL="Created Date"
 COM_CONTENT_FIELD_FEATURED_DESC="Assign the article to the featured blog layout."
+COM_CONTENT_FIELD_FEATURED_DOWN_DESC="An optional date to Finish being Featured the article."
```
Can we use end instead of Finish please
> +COM_CONTENT_FIELD_FEATURED_DOWN_DESC="An optional date to Finish being Featured the article."

+COM_CONTENT_FIELD_FEATURED_DOWN_DESC="An optional date to end the article being Featured."

> +COM_CONTENT_FIELD_FEATURED_DOWN_LABEL="Finish being Featured"

+COM_CONTENT_FIELD_FEATURED_DOWN_LABEL="End Featured"

> +COM_CONTENT_FIELD_FEATURED_UP_DESC="An optional date to Start being Featured the article."

+COM_CONTENT_FIELD_FEATURED_UP_DESC="An optional date to start the article being featured."

> +COM_CONTENT_FIELD_FEATURED_UP_LABEL="Start being Featured"

+COM_CONTENT_FIELD_FEATURED_UP_LABEL="Start Featured"

eshiol replied:
ok

brianteeman commented on libraries/fof/table/table.php:
```diff
@@ -3788,6 +3788,8 @@ public function checkContentType($alias = null)
 							"core_params"          => $this->getUcmCoreAlias('params'),
 							"core_featured"        => $this->getUcmCoreAlias('featured'),
+							"core_featured_up"     => $this->getUcmCoreAlias('featured_up'),
+							"core_featured_down"   => $this->getUcmCoreAlias('featured_down'),
```
Not sure why you are changing this library. Pretty sure its not used for what you are doing and we dont usually modify upstream libraries

eshiol replied:
Ok. How can I revert these changes?

brianteeman replied:
just delete them from your branch

eshiol committed:
fixes language strings

eshiol committed:
Merge branch 'Schedule-featured-articles-#13596' of https://github.com/eshiol/joomla-cms into Schedule-featured-articles-#13596

brianteeman commented on administrator/language/en-GB/en-GB.com_content.ini:
```diff
@@ -58,6 +58,14 @@ COM_CONTENT_FIELD_CREATED_BY_LABEL="Created By"
+COM_CONTENT_FIELD_FEATURED_DOWN_DESC="An optional date to Finish being Featured the article."
+COM_CONTENT_FIELD_FEATURED_DOWN_LABEL="Finish being Featured"
+COM_CONTENT_FIELD_FEATURED_UP_DESC="An optional date to Start being Featured the article."
+COM_CONTENT_FIELD_FEATURED_UP_LABEL="Start being Featured"
```
you have added my proposed strings but you didnt delete yours so now we have duplicates

eshiol committed:
fixes language strings

eshiol committed:
revert libraries/fof/table/table.php

brianteeman commented:
you have deleted the fof/table/table.php file instead of reverting the changes

photodude commented:
@alikon the `#__content_frontpage` table is id sorting, it doesn't control featured status. With the current database structure, the `#__content` table should be correct.

alikon commented:
from my data model POV is wrong, these fields  are attributes of the featured and not for the content entity

photodude commented:
@eshiol 
- does this handle the case where `featured_up` is `now` and `featured === 0`? 
 - what is the expected behavior in that case?
- does this handle the case where `featured_down` is `now` and `featured === 1`? 
 - what is the expected behavior in that case?

photodude commented:
@alikon I don't disagree. It may hit even a point where featured should be in its own table. I was just commenting on the current Joomla core structure which uses the `#__content` table to control featured items.

photodude commented:
@eshiol Please use the PR template for your PR. The template gives you places for the details needed for adding a feature, for how testing should proceed and more.  

```md
Pull Request for Issue #13596.

Summary of Changes

Testing Instructions

Expected result

Actual result

Documentation Changes Required
```

mbabker committed:
Deleted files update

mbabker committed:
Merge branch 'staging' of github.com:joomla/joomla-cms into 3.8-dev

eshiol commented:
Pull Request for Issue #13596.

### Summary of Changes
Added two fields in the article editor `featured_up` (Start featured) and `featured_down` (End featured)
An article is featured if `featured` is yes and (`featured_up` is null or is less than or equals to now) and (`featured_down` is null or is greater than or equal to now)

### Testing Instructions
create 5 articles:
1) not featured
2) featured, featured_up = null, featured_down = null
3) featured, featured_up = tomorrow, featured_down = null
4) featured, featured_up = null, featured_down = yesterday
5) featured, featured_up = yesterday, featured_down = tomorrow

Create an `Articles » Featured Articles` menu item
Create a `News Flash` module and set the option `Featured Articles` to `Only show Featured Articles`.

### Expected result
Only the articles 2 and 5 will be shown in the `Articles » Featured Articles` menu item and in the `News Flash` module.

### Actual result

### Documentation Changes Required

franz-wohlkoenig commented:
@eshiol can you please adapt your first Comment with your Comment above? So Tester can read what and how to test and don't have to scroll down.

eshiol committed:
Revert "revert libraries/fof/table/table.php"

This reverts commit 6e0b2db7194151e8397ff3de3aa45cb394146674.

eshiol committed:
revert fof changes

eshiol commented:
What about adding featured dates in the toggle featured tooltip?
![image](https://user-images.githubusercontent.com/12718836/27853399-c8a56ce6-6162-11e7-8bba-8fbb9689f951.png)

eshiol committed:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into Schedule-featured-articles-#13596

eshiol committed:
Check the featured dates

Check the featured down date is not earlier than featured up.

zero-24 commented on administrator/components/com_admin/sql/updates/mysql/3.7.3-2017-07-01.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE `#__content` ADD COLUMN `featured_up` datetime NOT NULL DEFAULT '0000-00-00 00:00:00';
```
IIRC this default date does not work anymore on mysql 5.7?

brianteeman replied:
that is true it has issues (outside of Joomla) BUT we are still using it everywhere so it should stay as proposed for this PR

eshiol replied:
The default SQL mode in MySQL 5.7 includes these modes: NO_ZERO_DATE and NO_ZERO_IN_DATE.
The NO_ZERO_DATE and NO_ZERO_IN_DATE modes were added in MySQL 5.7.8.

You can use `SET sql_mode='';` before executing the ALTER TABLE query or you can disable these modes globally
```
SET GLOBAL sql_mode=(SELECT REPLACE(@@sql_mode,'NO_ZERO_IN_DATE',''));
SET GLOBAL sql_mode=(SELECT REPLACE(@@sql_mode,'NO_ZERO_DATE',''));

```

tonypartridge commented:
I agree with @brianteeman until a fix is done MySQL 5.7 this is correct.

On 5 Jul 2017, 19:30 +0100, Brian Teeman <notifications@github.com>, wrote:
> @brianteeman commented on this pull request.
> In administrator/components/com_admin/sql/updates/mysql/3.7.3-2017-07-01.sql:
> > @@ -0,0 +1,2 @@
> +ALTER TABLE `#__content` ADD COLUMN `featured_up` datetime NOT NULL DEFAULT '0000-00-00 00:00:00';
> that is true it has issues (outside of Joomla) BUT we are still using it everywhere so it should stay as proposed for this PR
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

eshiol committed:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into Schedule-featured-articles-#13596

eshiol committed:
Add featured_up and featured_down

eshiol committed:
Revert "Add featured_up and featured_down"

This reverts commit 1efc698abd1cb1f781b93045b75a3610c2d1eb22.

photodude commented:
@eshiol you will need to adjust the unit test for `JTableContentTest` with the specific test `testCheck`
so it doesn't throw `Undefined property: JTableContent::$featured_down` (I believe you just need to add your new table fields to the test expectations)

eshiol committed:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into Schedule-featured-articles-#13596

mbabker committed:
Merge branch 'staging' of github.com:joomla/joomla-cms into 3.8-dev

fastslack committed:
[3.8] Namespace Http classes (#16445)

* Namespace Http classes

* Fix class name

* Fix class namespace name

* Fix wrapper namespace name

* Fix transport filename

* Fix codestyle

* Fix codestyle

* Solve unit test for J4

* Fix BC break

* Ups

mbabker committed:
Correct logic to match names

mbabker committed:
Rename class to match Framework

mbabker committed:
Rename interface to match the Framework

eshiol committed:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into Schedule-featured-articles-#13596

eshiol committed:
fixes unit test

photodude commented:
Thanks @eshiol I think this PR is now in a testable state. I'm not sure when I will get to testing, I hope within the next week.

Quy commented on administrator/language/en-GB/en-GB.com_content.ini:
```diff
@@ -58,6 +58,10 @@ COM_CONTENT_FIELD_CREATED_BY_LABEL="Created By"
 COM_CONTENT_FIELD_CREATED_DESC="Created date."
 COM_CONTENT_FIELD_CREATED_LABEL="Created Date"
 COM_CONTENT_FIELD_FEATURED_DESC="Assign the article to the featured blog layout."
+COM_CONTENT_FIELD_FEATURED_DOWN_DESC="An optional date to end the article being Featured."
```
Make lower case `Featured`.

eshiol replied:
ok

Quy commented on components/com_content/models/articles.php:
```diff
@@ -219,6 +221,8 @@ protected function getListQuery()
 			{
 				$query->where('a.featured = 1');
 			}
+			$query->where('(a.featured_up = ' . $nullDate . ' OR a.featured_up <= ' . $nowDate . ')')
```
Should these 2 lines be after `$query->where('a.featured = 1');`?

eshiol replied:
I don't think so. The full plain query could be produced by this piece of code
```
		if ($this->getState('filter.frontpage'))
		{
			// filter only featured articles
			$query->where('a.featured = 1')
				->where('(a.featured_up = ' . $nullDate . ' OR a.featured_up <= ' . $nowDate . ')')
				->where('(a.featured_down = ' . $nullDate . ' OR a.featured_down >= ' . $nowDate . ')');
			// join #__content_frontpage for ordering
			if ($orderby_sec === 'front')
			{
				$query->join('INNER', '#__content_frontpage AS fp ON fp.content_id = a.id');
			}
		}
```		
Articles have a row in the #__content_frontpage only if a.featured = 1, so you can optimize the query
```
		if ($this->getState('filter.frontpage'))
		{
			if ($orderby_sec === 'front')
			{
				$query->join('INNER', '#__content_frontpage AS fp ON fp.content_id = a.id');
			}
			else
			{
				$query->where('a.featured = 1');
			}
			$query->where('(a.featured_up = ' . $nullDate . ' OR a.featured_up <= ' . $nowDate . ')')
				->where('(a.featured_down = ' . $nullDate . ' OR a.featured_down >= ' . $nowDate . ')');
		}
```

Quy commented on tests/unit/schema/ddl.sql:
```diff
@@ -102,6 +102,8 @@ CREATE TABLE `jos_content` (
   `hits` INTEGER NOT NULL DEFAULT '0',
   `metadata` TEXT NOT NULL DEFAULT '',
   `featured` INTEGER NOT NULL DEFAULT '0',
+  `featured_up` TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
```
Type `TEXT`??

eshiol replied:
I used the same type of the other date fields
```
  `created` TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
  `modified` TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
  `checked_out_time` TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
  `publish_up` TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
  `publish_down` TEXT NOT NULL DEFAULT '0000-00-00 00:00:00',
```

photodude replied:
@eshiol I have opened PR #17031 to correct the test schema - changing `TEXT` or `TIMESTAMP` to `DATETIME`.  I have no idea why those fields in the test schema have been set to something other than `DATETIME` for so long.

csthomas commented on administrator/components/com_admin/sql/updates/sqlazure/3.7.4-2017-07-07.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE [#__content] ADD [featured_up] [datetime] NOT NULL DEFAULT '1900-01-01T00:00:00.000';
+ALTER TABLE [#__content] ADD [featured_down] [datetime] NOT NULL DEFAULT '1900-01-01T00:00:00.000';
```
Joomla does not need milliseconds. Take a look at https://github.com/joomla/joomla-cms/blob/staging/installation/sql/sqlazure/joomla.sql

You can use quotes instead brackets:
```sql
ALTER TABLE "#__content" ADD "featured_up" datetime2(0) NOT NULL DEFAULT '1900-01-01 00:00:00';
ALTER TABLE "#__content" ADD "featured_down" datetime2(0) NOT NULL DEFAULT '1900-01-01 00:00:00';
```

wojsmol replied:
@csthomas Plese test if this queries works on sqlazure if run using Joomla! database fixer - AFAIK no - long standing bug in database fixer.

eshiol replied:
Ok. I used the same standard used in the othe files like file 3.7.0-2016-08-29.sql
```
CREATE TABLE [#__fields] (
	[created_time] [datetime] NOT NULL DEFAULT '1900-01-01T00:00:00.000',
	[modified_time] [datetime] NOT NULL DEFAULT '1900-01-01T00:00:00.000',
```

wojsmol replied:
```sql
ALTER TABLE "#__content" ADD COLUMN "featured_up" datetime2(0) NOT NULL DEFAULT '1900-01-01 00:00:00';
ALTER TABLE "#__content" ADD COLUMN "featured_down" datetime2(0) NOT NULL DEFAULT '1900-01-01 00:00:00';

```
Whill work if this is valid syntax - no a sqlazure expert . Bug affects all supported database AFAIK.

eshiol committed:
fixes minor bugs

eshiol committed:
uses datetime2 instead of datetime

eshiol committed:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into Schedule-featured-articles-#13596

eshiol committed:
fixes test schema

eshiol commented:
I think we need to add something like the published state (Published and is Current; Published, but has Expired; Published, but is Pending) to the featured state (Featured; Featured, but Stopped; Featured, but not yet Started). I would use the icon "expired" for the stopped status, and icon "pending" for the not yet started status). What do you think?

![image](https://user-images.githubusercontent.com/12718836/28008942-da73c3b4-6559-11e7-887a-11b491900717.png)

tonypartridge commented:
Thats a good idea, I’d go simpler with:

Featured
Featured End: 2017-07-09 10:19:07
Featured Ended:  2017-07-09 10:19:07

On 10 Jul 2017, 09:29 +0100, Helios Ciancio <notifications@github.com>, wrote:
> I think we need to add something like the published state (Published and is Current; Published, but has Expired; Published, but is Pending) to the featured state (Featured; Featured, but Stopped; Featured, but not yet Started). I would use the icon "expired" for the stopped status, and icon "pending" for the not yet started status). What do you think?
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

eshiol committed:
add dates to the tooltip

eshiol committed:
Correct types in testGetFields() to use DATETIME

eshiol committed:
fixes code style

eshiol committed:
fixes cast statements

eshiol committed:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into Schedule-featured-articles-#13596

eshiol committed:
Merge branch 'staging' into Schedule-featured-articles-#13596

laoneo committed:
Namespace Input and deprecate it in favor of the framework package (#16939)

* Apply patch #12532, move to the fw package classes

* Namespace input classes

* Change application classes

* Compile correct class name

* Restore unit test

mbabker committed:
Deleted files update

mbabker committed:
Merge branch 'staging' of github.com:joomla/joomla-cms into 3.8-dev

brianteeman committed:
Update string in fields (#17119)

It was a bit ambiguous so I updated the string to be more explicit

### Before

### After

brianteeman committed:
Deprecated notice (#17094)

Update the deprecated notice on an unused file to indicate that it will be removed in J4

frankmayer committed:
[Plugins/system/debug] Some improvements and cleanups in plugins/system/debug (#16674)

* Some improvements and cleanups in plugins/system/debug
- Make use of $this->db instead of JFactory::getDbo()();
- Added missing docblocks
- Type-safety
- replace unnecessary strlen() call with simple check
- CS

* A few more changes

* A few more changes

* Changes based on reviewer's comment

* cs

* cs

* cs

* cs

* cs

brianteeman committed:
Change vote plugin to disabled (#17106)

On new installs this changes the vote plugin to be disabled by default. It is not used that often and by disabling it we also benefit from a simpler article list view in the admin

brianteeman committed:
Banned text strings consistency (#17108)

The text used for the banned fields in com_contacts is different in the commponent options and the individual contact options. Irrelevant of if the information is needed or not we should be consistent. So on the basis that it is important to have the extra information I have updated the individual contact strings to match the global ones.

izharaazmi committed:
Refactor database drivers - push up desctructor (#10962)

* Push up the destructor from various database drivers to their parent abstract class JDatabaseDriver as it almost every time just does one thing viz disconnect(). Otherwise it can always be overridden.
Never `unset` class property. $this->connection  set to null in disconnect methods.
Nosql connect return is void not bool.

* Use version macro now

kernusr committed:
put multiple versions of the same  tag (#16583)

* put multiple versions of the same <meta> tag

* fixed code-style errors

* remove tabs

* cs: Remove tabs

frankmayer committed:
Work on com_finder (both admin and site) (#12254)

* Case mismatch in function call

* Elvis

* Fix nested positive ifs

* Fix static method invocation via ->

* " to '

* Remove unnecessary parentheses

* Merged isset()

* Inline redundant value

* Move disconnected code outside the foreach loop.

* Fixed Docblocks

* Replaced preg_replace() with str_replace()

* Replaced PHP4 type-casting

* Type-safe string comparisons

* More type-safe comparisons

* Removed unused call and variable.
Also Todo'd two things for someone to check

* Reverse preg_replace() replacement because there was a limit in preg and it might cause trouble...

* Codesniffer

* Fix integer

*  Changes proposed by @andrepereiradasilva and two more by me, to get function calls out of the loop or reduce calls inside the loop.

* Shouldn't this be written this way?

* Some extra work according to @andrepereiradasilva's notes

* Some more changes from @andrepereiradasilva's comments

* Some more changes from @andrepereiradasilva's comments

* Reverted removal of duplicate key entries, added TODO's

* Code formatting fixes

* More code formatting fixes

* Fix merge introduced problem.

* Additional changes according to reviewer's comments

* Additional changes according to reviewer's comments

* Changes, according to @andrepereiradasilva's suggestions

* Forgot to commit this one

asika32764 committed:
Add CSRF token for Ajax (#14952)

* Add CSRF token for Ajax

* Add CSRF token to hathor

* CS

* call jquery first

* Use server->get()

* Loaded once

* Add CSRF to core.js request method

* Use CSRF instead Csrf

* Use scriptOption

* Fix test

* MISC update

* Use $name variable in JS

* Use string as script option directly and fix tests

brianteeman commented:
Can you look at resolving the conflicts so that this can be tested please.

eshiol committed:
Merge remote-tracking branch 'refs/remotes/joomla/3.8-dev' into Schedule-featured-articles-#13596

zero-24 commented:
@eshiol it looks like you have many not needed files included here?

eshiol commented:
I merged the remote-tracking branch 'refs/remotes/joomla/3.8-dev', my mistake. But now, I'm not able to revert it.

eshiol committed:
Revert "Merge remote-tracking branch 'refs/remotes/joomla/3.8-dev' into Schedule-featured-articles-#13596"

This reverts commit 956a9cdae01855e65a9f7c84fe4579c95fcb3e85, reversing
changes made to 92a92a42d6a548d617ec64a4c2107dbeef496424.

photodude commented:
@eshiol if `git revert 956a9cd` won't allow you to revert you will likely have to figure out how to redo the source work. (there is probably some way to create a new branch and cherry pick commit 09f5cd5 then overwrite the broken branch in this PR, but I'm not up on all of the pr "rescue" techniques for fixing mistakes and broken branches.)

tonypartridge commented:
Just do a git reset --hard upstream/staging if you want to make the branch
the same as the staging branch

-- 
Tony Partridge

From: Walt Sorensen <notifications@github.com> <notifications@github.com>
Reply: joomla/joomla-cms
<reply@reply.github.com>
<reply@reply.github.com>
Date: 22 August 2017 at 04:31:36
To: joomla/joomla-cms <joomla-cms@noreply.github.com>
<joomla-cms@noreply.github.com>
CC: Tony Partridge <admin@xtech.im> <admin@xtech.im>, Mention
<mention@noreply.github.com> <mention@noreply.github.com>
Subject:  Re: [joomla/joomla-cms] New Feature: Schedule featured articles
#13596 (#16960)

@eshiol <https://github.com/eshiol> if git revert 956a9cd won't allow you
> to revert you will likely have to figure out how to redo the source work.
> (there is probably some way to create a new branch and cherry pick commit
> 09f5cd5
> <https://github.com/joomla/joomla-cms/commit/09f5cd5dd1e199cb8a4df4dcda2d66c3e4502f6e>
> then overwrite the broken branch in this PR, but I'm not up on all of the
> pr "rescue" techniques for fixing mistakes and broken branches.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/16960#issuecomment-323909534>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABVgloes6B-Mu9YvlwJDT9WeXs0136zzks5sakuXgaJpZM4ONEhM>
> .
>

eshiol commented:
Broken PR. I'll open a new one. Sorry.

mijalis commented:
I just installed 3.8.0 and this feature is not yet available in the backend.

brianteeman commented:
@mijalis As stated above this was closed by the submitter and is awaiting a new pull request
